src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+and composition with linear maps `f`, `Q.comp f x = Q (f x)`.
+
+## Main definitions
+
+ * `quadratic_form.associated`: associated bilinear form
+ * `quadratic_form.pos_def`: positive definite quadratic forms
+ * `quadratic_form.discr`: discriminant of a quadratic form
+
+## Main statements
+
+ * `quadratic_form.associated_left_inverse`,
+ * `quadratic_form.associated_right_inverse`: in a commutative ring where 2 has
+  an inverse, there is a correspondence between quadratic forms and symmetric
+  bilinear forms
+
+## References
+
+ * https://en.wikipedia.org/wiki/Quadratic_form
+ * https://en.wikipedia.org/wiki/Discriminant#Quadratic_forms
+
+## Tags
+
+quadratic form, homogeneous polynomial, quadratic polynomial
+-/
+
+universes u v w
+variables {R : Type u} {M : Type v} [add_comm_group M] [ring R]
+variables {R₁ : Type u} [comm_ring R₁] [module R₁ M]
+
+/-- Up to a factor 2, `Q.polar` is the associated bilinear form for a quadratic form `Q`. -/
+def quadratic_form.polar (f : M → R) (x y : M) :=
+f (x + y) - f x - f y
+
+variables [module R M]
+
+open quadratic_form
+/-- A quadratic form over a module. -/
+structure quadratic_form (R : Type u) (M : Type v) [ring R] [add_comm_group M] [module R M] :=
+(map : M → R)
+(map_smul : ∀ (a : R) (x : M), map (a • x) = a * a * map x)
+(polar_add_left : ∀ (x x' y : M), polar map (x + x') y = polar map x y + polar map x' y)
+(polar_smul_left : ∀ (a : R) (x y : M), polar map (a • x) y = a • polar map x y)
+(polar_add_right : ∀ (x y y' : M), polar map x (y + y') = polar map x y + polar map x y')
+(polar_smul_right : ∀ (a : R) (x y : M), polar map x (a • y) = a • polar map x y)
+
+namespace quadratic_form
+
+variables {Q : quadratic_form R M}
+
+instance : has_coe_to_fun (quadratic_form R M) :=
+⟨_, λ B, B.map⟩
+
+/-- The `simp` normal form for a quadratic form is `coe_fn`, not `map`. -/
+@[simp] lemma map_eq_apply : Q.map = ⇑ Q := rfl
+
+lemma apply_smul (a : R) (x : M) : Q (a • x) = a * a * Q x := Q.map_smul a x
+
+lemma apply_add_self (x : M) : Q (x + x) = 4 * Q x :=
+by { rw [←one_smul R x, ←add_smul, apply_smul], norm_num }
+
+lemma apply_zero : Q 0 = 0 :=
+by rw [←@zero_smul R _ _ _ _ (0 : M), apply_smul, zero_mul, zero_mul]
+
+lemma apply_neg (x : M) : Q (-x) = Q x :=
+by rw [←@neg_one_smul R _ _ _ _ x, apply_smul, neg_one_mul, neg_neg, one_mul]
+
+lemma apply_sub (x y : M) : Q (x - y) = Q (y - x) :=
+by rw [←neg_sub, apply_neg]
+
+variable {Q' : quadratic_form R M}
+@[ext] lemma ext (H : ∀ (x : M), Q x = Q' x) : Q = Q' :=
+by { cases Q, cases Q', congr, funext, apply H }
+
+instance : has_zero (quadratic_form R M) :=
+⟨ { map := λ x, 0,
+    map_smul := λ a x, by simp,
+    polar_add_left := λ x x' y, by simp [polar],
+    polar_smul_left := λ a x y, by simp [polar],
+    polar_add_right := λ x y y', by simp [polar],
+    polar_smul_right := λ a x y, by simp [polar] } ⟩
+
+instance : inhabited (quadratic_form R M) := ⟨0⟩
+
+instance : has_scalar R₁ (quadratic_form R₁ M) :=
```

Comment:
What is the reason for the `R₁` as opposed to `R`?

Reply:
I've added a notation section in the module docs to explain the distinction between `R` and `R₁`. In this specific instance: `R₁` is commutative, which is needed for `map_smul : Q (a * b * x) = b^2 * Q (a * x)`

Reply:
Ok, thanks!